e2e/performance/app.test.js

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+'use strict';
+
+const env = require('../modules/environment');
+const {asyncSleep} = require('../modules/utils');
+
+describe('startup/app', function desc() {
+    this.timeout(30000);
+    this.slow(3000);
+
+    beforeEach(async () => {
+        env.createTestUserDataDir();
+        env.cleanTestConfig();
+        this.app = await env.getApp();
+    });
+
+    afterEach(async () => {
+        if (this.app) {
+            await this.app.close();
+        }
+        await env.clearElectronInstances();
+    });
+
+    it('should show the welcome screen modal when no servers exist', async () => {
+        if (this.app) {
+            await this.app.close();
+        }
+        await env.clearElectronInstances();
+        env.createTestUserDataDir();
+        env.cleanTestConfig();
+        this.app = await env.getApp();
+        await asyncSleep(500);
+        const welcomeScreenModal = this.app.windows().find((window) => window.url().includes('welcomeScreen'));
+        const modalButton = await welcomeScreenModal.innerText('.WelcomeScreen .WelcomeScreen__button');
+        modalButton.should.equal('Get Started');
```

Comment:
@tboulis 

An example of running the test multiple times and measure the results (based on the previous test that was in place in this draft, but could be applied to this example):

```js
    const measurePerformance = async (beforeRun, test, afterRun, numberOfRuns, durationThreshold) => {
        const testRuns = [];
        for (let i = 0; i < numberOfRuns; i++) {
            testRuns.push((async () => {
                await beforeRun();

                const t0 = performance.now();
                await test();
                const t1 = performance.now();

                await afterRun();

                return t1 - t0;
            })());
        }

        const durations = await Promise.all(testRuns);
        const avgDuration = durations.reduce((a, b) => a + b, 0) / durations.length;
        avgDuration.should.lessThanOrEqual(durationThreshold);
    };

    describe('Open settings window', () => {
        it('Should happen within expected duration', async () => {
            const expected = (process.platform === 'win32' || process.platform === 'linux');
            let settingsWindow;

            await measurePerformance(
                async () => {
                    this.app.evaluate(({ipcMain}, showWindow) => {
                        ipcMain.emit(showWindow);
                    }, SHOW_SETTINGS_WINDOW);
                },
                async () => {
                    settingsWindow = await this.app.waitForEvent('window', {
                        predicate: (window) => window.url().includes('settings'),
                    });
                    await settingsWindow.waitForSelector('.settingsPage.container');
                    await settingsWindow.waitForSelector('#inputAutoStart', {state: expected ? 'attached' : 'detached'});
                },
                async () => {
                    await settingsWindow.close();
                },
                10,
                500,
            );
        });
    });
```

The above uses `performance.now` which is available and runs the test multiple times, calculating an average duration of each run. The logic in that sense can be easily extracted in a helper (similar to how web app has `measurePerformance`) that can be reused by anyone. The pros of the approach is that we have access to the individual runs (so in the future we can do more advanced stuff, like having a deviation/percentiles of the runs instead of just averages, etc).

Reply:
Yep, I ran it (I don't know if the logic is fully correct, it's a quick thing I wrote). It runs and calculates 10 durations and averages them out.

Reply:
I don't have a particular preference (I'm 0/5) for one over the other approach (as I mentioned in the other comment where you suggest the loop on `it`). If you believe looping the test and adjusting the `checkReport` is a better approach, it's perfectly fine by me :)

Regarding the code, I can commit it (if you prefer the example I mentioned over the approach you suggest with `slow`), otherwise feel free to continue with the `slow` and looping over the tests approach if you feel it's more robust. Both accomplish the same purpose, so again 0/5.

Reply:
@tboulis with the multiple tests (looping over it) that you mention, perhaps the results may be more reliable. I've played around with the idea running this:

```js
    describe('Open settings window', () => {
        const durations = [];

        for (let i = 0; i < 3; i++) {
            it('Should happen within expected duration', async () => {
                const expected = (process.platform === 'win32' || process.platform === 'linux');

                this.app.evaluate(({ipcMain}, showWindow) => {
                    ipcMain.emit(showWindow);
                }, SHOW_SETTINGS_WINDOW);

                const t0 = performance.now();
                const settingsWindow = await this.app.waitForEvent('window', {
                    predicate: (window) => window.url().includes('settings'),
                });
                await settingsWindow.waitForSelector('.settingsPage.container');
                await settingsWindow.waitForSelector('#inputAutoStart', {state: expected ? 'attached' : 'detached'});
                const t1 = performance.now();

                durations.push(t1 - t0);
                await settingsWindow.close();

                if (durations.length === 3) {
                    const avgDuration = durations.reduce((a, b) => a + b, 0) / durations.length;
                    avgDuration.should.lessThanOrEqual(300);
                }
            });
        }
    });
```

And it runs one test after the other etc, so it works quite well as well (I have no idea if this can be abstracted in a similar way to a measurePerformance thing for others to use. Leaving that here to continue the discussion on the different approaches :)